### PR TITLE
[Experimental] Product Filters: add layout and block spacing support

### LINKED ISSUE #48022: Product Filters Redesign > Wrapper Block Settings: Add the "Block Spacing" setting
## Description
The "Product Filters" wrapper block contains a setting that allows merchants to set the spacing between the inner blocks. That can be a little bit tricky since each inner block will have to respect that setting on the Editor and on the front end.

## Expected Behavior:

- When a merchant edits the block spacing, that should be reflected on the frontend.
- When a merchant edits the block spacing, that should be reflected on the editor.

## Reference images

| Product FIlters Wrapper Block Settings |
|--------|
| image |

### PR DESCRIPTION
**Pull request by dinhtungdu**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR adds the Block Spacing setting to the Product Filters block by adding `blockGap` support. `layout` support is required for block spacing to work, so it was added in this PR too. The layout setting is based on the following design:

image

Closes #48022.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

This PR should be tested by developers only.

0. Ensure the experimental blocks are enabled:
    - If you use the WooCommerce Beta Tester Plugin, you can enable the experimental blocks in Settings > WCA Test Helper > Features. Toggle the `experimental-blocks` option.
    - If you build this PR locally, build the branch with `pnpm --filter='@woocommerce/plugin-woocommerce' watch:build` command to have experimental blocks available.
2. Edit Block Spacing in the Editor:
    - Go to Site Editor > Templates > Product Catalog, add the "Product Filters (Experimental)" wrapper block to the template.
    - See Layout settings and Block Spacing settings in the sidebar.
    - Adjust the block spacing setting.
    - Verify that the spacing changes are immediately visible in the editor.
1. Verify Spacing on Frontend:
    - Save the changes and view the page or post on the frontend.
    - Confirm that the spacing adjustments made in the editor are accurately reflected on the frontend.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

**Review by nefeline — COMMENTED** [review]
Thanks for working on this, Tung!

A few recommendations and questions:

- Could you please update the test instructions to ensure that whoever is testing this PR, has the "experimental blocks" feature enabled in the WooCommerce Beta Tester plugin? If they don't, the **Product Filters (Experimental)** block won't be available for insertion in the Product Catalog.

Here are a few screenshots of the results I got while testing this PR, both on the editor and on the frontend:

Screenshot 2024-06-13 at 14 22 32

Screenshot 2024-06-13 at 14 22 22


The Block Spacing setting works as expected, but to confirm: is this the expected outcome in terms of design?

**Comment by dinhtungdu** [comment]
@nefeline thank you for your review, I updated the testing instructions mentioning experimental blocks. About the design, I changed the default vertical align to `top`, so the horizontal layout should look better now.

image

I had a conversation with @jarekmorawski on wrapping, he said that we want wrapping to support smaller screens. It also looks better with top alignment, but I don't like the long attribute list. What do you think Jarek? Happy to follow any direction here to make horizontal layout look good by default.

image

**Comment by jarekmorawski** [comment]
> About the design, I changed the default vertical align to top, so the horizontal layout should look better now.

Thanks, @dinhtungdu. I confirm this should be the default setting. 🙏 

> It also looks better with top alignment, but I don't like the long attribute list. What do you think Jarek? Happy to follow any direction here to make horizontal layout look good by default.

Honestly, there's no way to make it look good without introducing dropdowns or accordions (currently WIP). We can, however, offer truncation if the list of attributes grows too long. I planned to make it a future improvement to limit the scope for the MVP, but feel free to take this on if you have space.

> If the attribute has more than 15 terms, we add a “Show more” button. It has the same styling as the “Clear” button. Users can’t click it in the canvas, but it will allow shoppers to expand the list of available terms on the frontend.

image

**Review by thealexandrelara — COMMENTED** [review]
Thank you for working on this @dinhtungdu. I just tested and here are the results:

✅ Merchants can justify items to the left, right, center or stretch items.
✅ Merchants can change the inner blocks orientation from vertical to horizontal.
⚠️ When the orientation is set to horizontal, the "Active" filters heading is not aligned to the top like the others:
CleanShot 2024-06-18 at 14 53 50@2x
⚠️ Since you were unable to join our last team meetup, I believe you're not aware of one of the things we discussed about always including E2E tests in our PRs. Would you mind including some E2E tests? For example, one test suggestion would be to use [CSS to match elements based on layout](https://playwright.dev/docs/other-locators#css-matching-elements-based-on-layout) to detect if the blocks appear horizontally when the orientation is set to horizontal.

**Comment by dinhtungdu** [comment]
> When the orientation is set to horizontal, the "Active" filters heading is not aligned to the top like the others:

As I mentioned [above](https://github.com/woocommerce/woocommerce/pull/48429#issuecomment-2167435040), in the new design, the Active Filters shouldn't have a title by default. I didn't touch that block in this PR because it's out of scope.

> always including E2E tests in our PRs

Love that. I will add some for this PR.

**Comment by dinhtungdu** [comment]
@thealexandrelara I add some tests to verify new settings. This PR is ready for another review, can you please give it another look?

File: plugins/woocommerce-blocks/tests/e2e/tests/product-filters/product-filters.block_theme.spec.ts [hunk 1]
```diff
@@ -158,5 +162,148 @@ test.describe( `${ blockData.name }`, () => {
...
+		await layoutSettings.getByLabel( 'Horizontal' ).click();
+		await expect(
+			layoutSettings.getByLabel( 'Stretch items' )
+		).toBeHidden();
+		await expect(
+			layoutSettings.getByLabel( 'Space between items' )
+		).toBeVisible();
+		await expect(
+			block.locator( blockData.selectors.editor.layoutWrapper )
+		).toHaveCSS( 'flex-direction', 'row' );
```
Comment (dinhtungdu) [inline plugins/woocommerce-blocks/tests/e2e/tests/product-filters/product-filters.block_theme.spec.ts]: I'm not happy with CSS locator and assertion, but it seems to be the only way to test these settings AFAIK. I'm open to any suggestions.

Comment (thealexandrelara) [reply plugins/woocommerce-blocks/tests/e2e/tests/product-filters/product-filters.block_theme.spec.ts]: One option would be to use css locators based on layout position, like in these [examples](https://playwright.dev/docs/other-locators#css-matching-elements-based-on-layout):

```
// Fill an input to the right of "Username".
await page.locator('input:right-of(:text("Username"))').fill('value');

// Click a button near the promo card.
await page.locator('button:near(.promo-card)').click();

// Click the radio input in the list closest to the "Label 3".
await page.locator('[type=radio]:left-of(:text("Label 3"))').first().click();
```

**Review by thealexandrelara — APPROVED** [review]
Great work! I left a comment regarding the CSS locators in Playwright, but it is not a blocker for this PR, so I'm pre-approving it!

File: plugins/woocommerce-blocks/tests/e2e/tests/product-filters/product-filters.block_theme.spec.ts (hunk 1 above)
Comment (dinhtungdu) [reply plugins/woocommerce-blocks/tests/e2e/tests/product-filters/product-filters.block_theme.spec.ts]: Ah, I checked your [suggestion](https://github.com/woocommerce/woocommerce/pull/48429#pullrequestreview-2126170020) but I haven't gotten it until now. We can check if the text Status which is right of Price is visible or not to verify the layout direction. Thank you, I will try it in this PR.

About the block spacing tests, do you have any alternative?

**Pull request merged**